src/shogun/clustering/GMM.cpp

```diff
@@ -523,113 +510,117 @@ void CGMM::max_likelihood(SGMatrix<float64_t> alpha, float64_t min_cov)
 
 	float64_t alpha_sum;
 	float64_t alpha_sum_sum=0;
-	float64_t* mean_sum;
-	float64_t* cov_sum=NULL;
 
 	for (int32_t i=0; i<alpha.num_cols; i++)
 	{
 		alpha_sum=0;
-		mean_sum=SG_MALLOC(float64_t, num_dim);
-		memset(mean_sum, 0, num_dim*sizeof(float64_t));
+		SGVector<float64_t> mean_sum(num_dim);
+		linalg::zero(mean_sum);
 
 		for (int32_t j=0; j<alpha.num_rows; j++)
 		{
 			alpha_sum+=alpha.matrix[j*alpha.num_cols+i];
 			SGVector<float64_t> v=dotdata->get_computed_dot_feature_vector(j);
-			SGVector<float64_t>::add(mean_sum, alpha.matrix[j*alpha.num_cols+i], v.vector, 1, mean_sum, v.vlen);
+			linalg::add(
+			    v, mean_sum, mean_sum, alpha.matrix[j * alpha.num_cols + i],
+			    1.0);
 		}
 
-		for (int32_t j=0; j<num_dim; j++)
-			mean_sum[j]/=alpha_sum;
+		linalg::scale(mean_sum, mean_sum, 1.0 / alpha_sum);
+
+		m_components[i]->set_mean(mean_sum);
 
-		m_components[i]->set_mean(SGVector<float64_t>(mean_sum, num_dim));
+		SGMatrix<float64_t> cov_sum;
 
 		ECovType cov_type=m_components[i]->get_cov_type();
 
 		if (cov_type==FULL)
 		{
-			cov_sum=SG_MALLOC(float64_t, num_dim*num_dim);
-			memset(cov_sum, 0, num_dim*num_dim*sizeof(float64_t));
+			cov_sum = SGMatrix<float64_t>(num_dim, num_dim);
+			linalg::zero(cov_sum);
 		}
 		else if(cov_type==DIAG)
 		{
-			cov_sum=SG_MALLOC(float64_t, num_dim);
-			memset(cov_sum, 0, num_dim*sizeof(float64_t));
+			cov_sum = SGMatrix<float64_t>(1, num_dim);
+			linalg::zero(cov_sum);
 		}
 		else if(cov_type==SPHERICAL)
 		{
-			cov_sum=SG_MALLOC(float64_t, 1);
-			cov_sum[0]=0;
+			cov_sum = SGMatrix<float64_t>(1, 1);
+			linalg::zero(cov_sum);
 		}
 
 		for (int32_t j=0; j<alpha.num_rows; j++)
 		{
 			SGVector<float64_t> v=dotdata->get_computed_dot_feature_vector(j);
-			SGVector<float64_t>::add(v.vector, 1, v.vector, -1, mean_sum, v.vlen);
 
+			linalg::add(v, mean_sum, v, 1.0, -1.0);
 			switch (cov_type)
 			{
 				case FULL:
-					cblas_dger(CblasRowMajor, num_dim, num_dim, alpha.matrix[j*alpha.num_cols+i], v.vector, 1, v.vector,
-								 1, (double*) cov_sum, num_dim);
-
-					break;
-				case DIAG:
-					for (int32_t k=0; k<num_dim; k++)
-						cov_sum[k]+=v.vector[k]*v.vector[k]*alpha.matrix[j*alpha.num_cols+i];
-
-					break;
-				case SPHERICAL:
-					float64_t temp=0;
-
-					for (int32_t k=0; k<num_dim; k++)
-						temp+=v.vector[k]*v.vector[k];
-
-					cov_sum[0]+=temp*alpha.matrix[j*alpha.num_cols+i];
-					break;
+				    cblas_dger(
+				        CblasRowMajor, num_dim, num_dim,
+				        alpha.matrix[j * alpha.num_cols + i], v.vector, 1,
+				        v.vector, 1, (double*)cov_sum.matrix, num_dim);
+
+				    break;
+			    case DIAG:
+				    for (int32_t k = 0; k < num_dim; k++)
```

Comment:
` alpha.matrix[j * alpha.num_cols + i] ` is a element rather than a vector or matrix. So I think it's something like vector^{2}*R(R is a real number) rather than matrix multiplication. Does it make sense to u? :)